File: archive_experimental/watershed_ml_discovery.pl
#!/usr/bin/env python3
"""
watershed_ml_discovery.py
Unsupervised clustering of watershed chemistry data using PCA + KMeans.
Author: Ashley Reagan Mitchell
"""

import pandas as pd
import numpy as np
import os
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import plotly.express as px

# =========================================================
# Paths
# =========================================================
base_dir = os.path.dirname(os.path.abspath(__file__))
data_dir = os.path.join(base_dir, "data_outputs")

# Automatically find all *_latest.csv chemistry files
chem_files = [f for f in os.listdir(data_dir) if f.endswith("_latest.csv")]

if not chem_files:
    raise FileNotFoundError("No *_latest.csv chemistry files found in data_outputs/")

print(f"🧩 Found {len(chem_files)} chemistry files")
for f in chem_files:
    print("   ├─", f)

# =========================================================
# Merge Chemistry Files
# =========================================================
dfs = []
for f in chem_files:
    path = os.path.join(data_dir, f)
    df = pd.read_csv(path)
    if "MonitoringLocationIdentifier" in df.columns and "ResultMeasureValue" in df.columns:
        df = df.rename(columns={"MonitoringLocationIdentifier": "SiteID"})
        df["CharacteristicName"] = f.replace("_latest.csv", "").title()
        dfs.append(df[["SiteID", "CharacteristicName", "ResultMeasureValue", "LatitudeMeasure", "LongitudeMeasure", "HUCEightDigitCode", "HUCTwelveDigitCode"]])

if not dfs:
    raise RuntimeError("No valid data found in chemistry files.")

data = pd.concat(dfs, ignore_index=True)

# =========================================================
# Clean & Aggregate
# =========================================================
print("🧹 Cleaning and aggregating data...")
data = data.dropna(subset=["ResultMeasureValue"])
data["ResultMeasureValue"] = pd.to_numeric(data["ResultMeasureValue"], errors="coerce")
data = data.dropna(subset=["ResultMeasureValue"])
data = data.rename(columns={"HUCTwelveDigitCode": "HUC12"})

# Average by HUC12 and characteristic
chem_pivot = (
    data.groupby(["HUC12", "CharacteristicName"])["ResultMeasureValue"]
    .mean()
    .unstack()
    .reset_index()
)

chem_pivot = chem_pivot.dropna(axis=0, how="any")  # drop incomplete rows

# =========================================================
# Normalize
# =========================================================
features = chem_pivot.drop(columns=["HUC12"])
scaler = StandardScaler()
X_scaled = scaler.fit_transform(features)

# =========================================================
# PCA
# =========================================================
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_scaled)

explained_var = np.sum(pca.explained_variance_ratio_) * 100
print(f"🔍 PCA completed (2 components explain {explained_var:.2f}% variance)")

# =========================================================
# KMeans Clustering
# =========================================================
k = 4  # default clusters
model = KMeans(n_clusters=k, random_state=42, n_init=10)
labels = model.fit_predict(X_pca)
sil_score = silhouette_score(X_pca, labels)
print(f"🤖 KMeans clustering done (k={k}, silhouette={sil_score:.3f})")

chem_pivot["Cluster"] = labels
chem_pivot["PC1"] = X_pca[:, 0]
chem_pivot["PC2"] = X_pca[:, 1]

# =========================================================
# Save Outputs
# =========================================================
os.makedirs(data_dir, exist_ok=True)
chem_pivot.to_csv(os.path.join(data_dir, "watershed_clusters.csv"), index=False)

cluster_summary = chem_pivot.groupby("Cluster").mean(numeric_only=True)
cluster_summary.to_csv(os.path.join(data_dir, "cluster_summary.csv"))

# =========================================================
# Plot
# =========================================================
fig = px.scatter(
    chem_pivot,
    x="PC1",
    y="PC2",
    color=chem_pivot["Cluster"].astype(str),
    hover_data=["HUC12"],
    title="Watershed Chemistry Clusters (PCA + KMeans)",
    color_discrete_sequence=px.colors.qualitative.Bold
)
fig.write_html(os.path.join(data_dir, "cluster_scatter.html"))

# =========================================================
# Done
# =========================================================
print(f"""
✅ Saved results:
   ├─ watershed_clusters.csv
   ├─ cluster_summary.csv
   ├─ cluster_scatter.html
📊 PCA variance explained: {explained_var:.2f}%
📈 Silhouette score: {sil_score:.3f}
""")

File: archive_experimental/watershed_ml_discovery.py
# ============================================================
# watershed_ml_discovery_v2.py
# Machine Learning Discovery for Watershed Chemistry Patterns
# ============================================================

import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import matplotlib.pyplot as plt
import geopandas as gpd
from pathlib import Path
import warnings

warnings.filterwarnings("ignore")

# ------------------------------------------------------------
# 1️⃣ Load data
# ------------------------------------------------------------
input_path = Path("data_outputs/PA_wq_chemistry_clean.parquet")
df = pd.read_parquet(input_path)
print(f"✅ Loaded {len(df):,} records from {input_path.name}")

# ------------------------------------------------------------
# 2️⃣ Select feature columns
# ------------------------------------------------------------
chem_keywords = [
    "fe", "iron", "mn", "al", "ph", "temp", "cond", "so4", "sulf",
    "flow", "delta", "annualom", "cost", "maint"
]

chem_cols = [
    c for c in df.columns
    if any(k in c.lower() for k in chem_keywords)
    and pd.api.types.is_numeric_dtype(df[c])
]

if len(chem_cols) < 5:
    print(f"⚠️ Only {len(chem_cols)} chemistry fields found — expanding to all numeric columns.")
    chem_cols = [c for c in df.columns if pd.api.types.is_numeric_dtype(df[c])]

# Drop rows missing all these values
X = df[chem_cols].dropna(how="all")

if len(X) < 500:
    print(f"⚠️ Only {len(X)} usable rows — clustering may be unstable.")

print(f"🧪 Using {len(chem_cols)} numeric features across {len(X):,} samples.")
print(", ".join(chem_cols))

# ------------------------------------------------------------
# 3️⃣ Scale and reduce dimensionality
# ------------------------------------------------------------
X_scaled = StandardScaler().fit_transform(X.fillna(0))

pca = PCA(n_components=0.9, svd_solver="full")
X_pca = pca.fit_transform(X_scaled)

print(f"📉 PCA reduced features → {X_pca.shape[1]} components (90% variance)")

# ------------------------------------------------------------
# 4️⃣ Auto-select best cluster count (2–8)
# ------------------------------------------------------------
best_k, best_score = 2, -1
for k in range(2, 9):
    km = KMeans(n_clusters=k, random_state=42, n_init=10)
    labels = km.fit_predict(X_pca)
    score = silhouette_score(X_pca, labels)
    if score > best_score:
        best_k, best_score = k, score

print(f"🏷️ Optimal cluster count: {best_k} (silhouette={best_score:.3f})")

kmeans = KMeans(n_clusters=best_k, random_state=42, n_init=20)
clusters = kmeans.fit_predict(X_pca)
df["Cluster"] = np.nan
df.loc[X.index, "Cluster"] = clusters

# ------------------------------------------------------------
# 5️⃣ Cluster summary
# ------------------------------------------------------------
summary = df.groupby("Cluster")[chem_cols].mean().round(2)
summary["Count"] = df.groupby("Cluster").size()
summary = summary.reset_index()

print("\n📊 Cluster summary (first 10 columns):")
print(summary.iloc[:, :10])

# ------------------------------------------------------------
# 6️⃣ Save outputs
# ------------------------------------------------------------
out_dir = Path("data_outputs")
csv_path = out_dir / "PA_wq_clusters_v2.csv"
summary_path = out_dir / "PA_wq_cluster_summary_v2.csv"
jpg_path = out_dir / "PA_wq_cluster_plot_v2.jpg"
geo_path = out_dir / "PA_wq_clusters_v2.geojson"

df.to_csv(csv_path, index=False)
summary.to_csv(summary_path, index=False)

# ------------------------------------------------------------
# 7️⃣ Save GeoJSON if geometry present
# ------------------------------------------------------------
geo_cols = [c for c in df.columns if c.lower() in ("geometry", "geom", "wkt")]
if geo_cols:
    gdf = gpd.GeoDataFrame(df.dropna(subset=geo_cols), geometry=geo_cols[0], crs="EPSG:4326")
    gdf.to_file(geo_path, driver="GeoJSON")
    print(f"🗺️ GeoJSON saved → {geo_path}")
else:
    print("⚠️ No geometry column found — skipping GeoJSON export.")

# ------------------------------------------------------------
# 8️⃣ PCA scatter plot
# ------------------------------------------------------------
plt.figure(figsize=(8, 6))
scatter = plt.scatter(
    X_pca[:, 0],
    X_pca[:, 1],
    c=clusters,
    cmap="viridis",
    s=10,
    alpha=0.6
)
plt.colorbar(scatter, label="Cluster")
plt.title("PCA Cluster Visualization – Watershed Chemistry")
plt.xlabel("PCA 1")
plt.ylabel("PCA 2")
plt.tight_layout()
plt.savefig(jpg_path, dpi=300)
plt.close()
print(f"🖼️ PCA plot saved → {jpg_path}")

print("\n✅ watershed_ml_discovery_v2 complete.")
print(f"Outputs saved in: {out_dir.resolve()}")